File: src/FlashcardParser/LineTabCounter.ts

/**
 * Countes the number of tabs a line starts with. First condifering 
 * tab characters, then looking for spaces.
 * @param line The string line.
 * @returns The number of tabs found.
 */
export function countTabs(line: string): number {
    let count = 0;
    let i = 0;

    while (i < line.length) {
        if (line[i] === '\t') {
            count++;
            i++;
        } else if (line[i] === ' ') {
            // Check if we have 4 consecutive spaces
            if (i + 3 < line.length &&
                line[i + 1] === ' ' &&
                line[i + 2] === ' ' &&
                line[i + 3] === ' ') {
                count++;
                i += 4; // Skip the 4 spaces
            } else {
                // Not 4 consecutive spaces, stop counting
                break;
            }
        } else {
            // Hit a non-whitespace character, stop counting
            break;
        }
    }

    return count;
}

File: service/flashcard_parser/line_tab_counter.py
def count_tabs(line: str) -> int:
    count = 0
    i = 0
    while i < len(line):
        if line[i] == '\t':
            count += 1
            i += 1
        elif line[i] == ' ':
            if i + 3 < len(line) and line[i + 1] == ' ' and line[i + 2] == ' ' and line[i + 3] == ' ':
                count += 1
                i += 4
            else:
                break
        else:
            break
    return count
